atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TopListFilter.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import com.google.common.base.Suppliers;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Contains the top N elements based on a score, updated eagerly within a memoization interval.
+ * During the evaluation of an update, if the top list is perceived to be not full (due to memoization),
+ * the current update element is considered to be in the top list.
+ */
+public final class TopListFilter<K> {
+    private final int maxSize;
+    private final Supplier<Set<K>> eligibleKeysSupplier;
+
+    private final ConcurrentMap<K, Double> entries = new ConcurrentHashMap<>();
+    private final Comparator<Map.Entry<K, Double>> reversedScoreEntryComparator =
+            Map.Entry.comparingByValue(Comparator.reverseOrder());
+
+    TopListFilter(int maxSize, Duration resetInterval) {
+        this.maxSize = maxSize;
+        this.eligibleKeysSupplier = Suppliers.memoizeWithExpiration(this::computeEligibleKeysInternal, resetInterval);
+    }
```

Comment:
Should this and the method be public? Alternatively, should the class have package visibility?

Reply:
No if it's not used by code that requires it here, I most likely just forgot to think about visibility here. Might be worth adding a question to the PR checklist for this?